contracts/iotube/TransferValidatorV3.sol

```diff
@@ -0,0 +1,135 @@
+pragma solidity <6.0 >=0.4.24;
+
+import "../lifecycle/Pausable.sol";
+
+interface IAllowlist {
+    function isAllowed(address) external view returns (bool);
+    function numOfActive() external view returns (uint256);
+}
+
+interface IMinter {
+    function mint(address, address, uint256) external returns(bool);
+    function transferOwnership(address) external;
+    function owner() external view returns(address);
+}
+
+interface IReceiver {
+    function onReceive(bytes32 id, address sender, address token, uint256 amount, bytes calldata payload) external;
+}
+
+contract TransferValidatorV3 is Pausable {
+    event Settled(bytes32 indexed key, address[] witnesses);
+    event ReceiverAdded(address receiver);
+    event ReceiverRemoved(address receiver);
+
+    mapping(bytes32 => uint256) public settles;
+    mapping(address => bool) public receivers;
+
+    IMinter[] public minters;
+    IAllowlist[] public tokenLists;
+    IAllowlist public witnessList;
+
+    constructor(IAllowlist _witnessList) public {
+        witnessList = _witnessList;
+    }
+
+    function generateKey(address cashier, address tokenAddr, uint256 index, address from, address to, uint256 amount) public view returns(bytes32) {
+        return keccak256(abi.encodePacked(address(this), cashier, tokenAddr, index, from, to, amount));
+    }
+
+    function submit(address cashier, address tokenAddr, uint256 index, address from, address to, uint256 amount, bytes memory signatures, bytes calldata payload) public whenNotPaused {
+        require(amount != 0, "amount cannot be zero");
+        require(to != address(0), "recipient cannot be zero");
+        require(signatures.length % 65 == 0, "invalid signature length");
+        bytes32 key = generateKey(cashier, tokenAddr, index, from, to, amount);
+        require(settles[key] == 0, "transfer has been settled");
+        for (uint256 it = 0; it < tokenLists.length; it++) {
+            if (tokenLists[it].isAllowed(tokenAddr)) {
+                uint256 numOfSignatures = signatures.length / 65;
+                address[] memory witnesses = new address[](numOfSignatures);
+                for (uint256 i = 0; i < numOfSignatures; i++) {
+                    address witness = recover(key, signatures, i * 65);
+                    require(witnessList.isAllowed(witness), "invalid signature");
+                    for (uint256 j = 0; j < i; j++) {
+                        require(witness != witnesses[j], "duplicate witness");
+                    }
+                    witnesses[i] = witness;
+                }
+                require(numOfSignatures * 3 > witnessList.numOfActive() * 2, "insufficient witnesses");
```

Comment:
Move L58 to above of L49 can be more efficiently when insufficient witnesses occur